### PR TITLE
PRODUCT-259 Implement persistent session management with connection reuse options

### PR DESCRIPTION
**Pull request by DrPyser**
## Summary
This PR refactors the REST client to maintain a persistent session across multiple calls, while adding configurable connection reuse and keep-alive header support. Previously, a new session was created for each request; now the client reuses the same session instance.

## Key Changes
- **Persistent Session Management**: The client now maintains a single `Session` instance (`self._session`) that is reused across calls instead of creating a new session for each `session()` call
- **Connection Reuse Configuration**: Added `connection_reuse` parameter (default: `True`) to control whether connections should be reused
  - When `False`: Sets `Connection: close` header
  - When `True`: Allows connection pooling and keep-alive headers
- **Keep-Alive Headers**: Added `keep_alive_timeout` and `keep_alive_max` parameters to configure HTTP keep-alive behavior
  - Headers are only set when `connection_reuse=True`
  - Supports both parameters independently or together
- **Dynamic Header Updates**: Modified `set_token()` and `tenant_uuid` setter to update headers on the existing session instead of requiring session recreation
- **Tenant UUID Property**: Converted `tenant_uuid` to a proper property with getter/setter that updates the session headers when changed

## Implementation Details
- Session creation is now lazy: the session is only created on first `session()` call
- The `_create_session()` method encapsulates session initialization logic
- Keep-alive header formatting supports combining timeout and max values (e.g., `"timeout=5, max=100"`)
- Updated test suite includes comprehensive coverage for session persistence, header management, and connection reuse scenarios
- Updated copyright year to 2026 in test file

https://claude.ai/code/session_011hCAeKaMZtN1JhVSenjk5a

---

> [!NOTE]
> **Medium Risk**
> Changes core HTTP client behavior for all Wazo services (connection lifecycle and header injection); well covered by tests but could affect threading or long-lived auth edge cases.
> 
> **Overview**
> **Reuses one `requests.Session` per client** instead of building a new session (and forcing `Connection: close`) on every `session()` call, so inter-service HTTP calls can use HTTP/1.1 keep-alive.
> 
> Session creation is **lazy and thread-safe** (`_session` + lock). **Token, tenant, and timeout** are applied on each request via wrapped `prepare_request` / `session.request` (caller headers override, case-insensitive), avoiding mutating shared session headers after `set_token()` or `tenant_uuid` changes.
> 
> Tests now assert persistence, dynamic token/tenant/timeout behavior, and an integration check that TCP connections are reused; README documents connection reuse.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 780d6109260a2daed73f18225034e5dab5f1eb4b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: wazo_lib_rest_client/tests/test_client.py [hunk 1]
```diff
@@ -117,6 +117,14 @@ def test_client_command_after_session_expiry(self):
 
         time.sleep(2)
 
+        # The client now keeps a persistent session, so its cookie jar
+        # survives across calls. This test's server stores its digest-auth
+        # nonce in a Flask session cookie that expires after 1s; resending
+        # the stale cookie breaks the re-auth handshake. Clearing the jar
+        # reproduces the original "fresh session per call" precondition so we
+        # still exercise recovery from server-side auth-session expiry.
```
Comment (DrPyser) [inline wazo_lib_rest_client/tests/test_client.py]: trim the comment into a more concise form (one sentence)

File: wazo_lib_rest_client/client.py [hunk 2]
```diff
@@ -93,8 +101,29 @@ def _load_plugins(self) -> None:
...
+        requests across threads; use one client per thread (or
+        ``connection_reuse=False``) in that case.
+        """
+        if self._session is None:
+            self._session = self._create_session()
+        return self._session
+
+    def _create_session(self) -> Session:
         session = Session()
-        session.headers = {'Connection': 'close'}
```
Comment (fblackburn1) [inline wazo_lib_rest_client/client.py:97]: This line was explicitly added to avoid to reuse session and lock a thread (which was limited to 10) for a service.
So this change depend of dynamic threading system AND some benchmark (can be done manually) to be sure threads are not exhausted

Moreover, since client python could be used on non-updated stack, there may be have something to think about default behavior (or communication)

Comment (DrPyser) [reply wazo_lib_rest_client/client.py:97]: yeah I know, hence why the client constructor can disable connection reuse. My plan is to disable connection reuse in the confd client, since that's where the problem occurred. Otherwise we could make it a per-service configuration parameter?

File: wazo_lib_rest_client/client.py [hunk 3]
```diff
@@ -93,13 +100,41 @@ def _load_plugins(self) -> None:
...
+        The session is created lazily on first use and reused for the
+        lifetime of the client so that HTTP connections are reused. A
+        ``requests.Session`` is not guaranteed thread-safe, so a single
+        client instance must not be shared for unsynchronized concurrent
+        requests across threads; use one client per thread (or
+        ``connection_reuse=False``) in that case.
+        """
+        if self._session is None:
+            self._session = self._create_session()
+        return self._session
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_lib_rest_client/client.py:115]: question is, should we allow cross-thread sharing of Session (is it threadsafe?), or should we make it thread-local?

if Session not threadsafe, for the change to be safe and drop-in, we would need to ensure transparent threadsafety through e.g. threadlocal Session store.

File: wazo_lib_rest_client/client.py [hunk 4]
```diff
@@ -93,13 +100,41 @@ def _load_plugins(self) -> None:
...
 
     def session(self) -> Session:
+        """Return the client's persistent ``requests.Session``.
+
+        The session is created lazily on first use and reused for the
+        lifetime of the client so that HTTP connections are reused. A
+        ``requests.Session`` is not guaranteed thread-safe, so a single
+        client instance must not be shared for unsynchronized concurrent
+        requests across threads; use one client per thread (or
+        ``connection_reuse=False``) in that case.
```
Comment (DrPyser) [inline wazo_lib_rest_client/client.py]: will want to trim this docstring

Comment (DrPyser) [reply wazo_lib_rest_client/client.py]: also is not in line with how clients are used currently (to review)

File: README.md [hunk 5]
```diff
@@ -69,6 +69,38 @@ client_1 = Client(https=True, verify_certificate=True)
...
+    host='localhost',
+    connection_reuse=True,
+    keep_alive_timeout=5,
+    keep_alive_max=100,
+)
+```
+
+Because the session is shared for the lifetime of the client, a single client
+instance is not safe for unsynchronized concurrent requests from multiple
+threads (a `requests.Session` is not guaranteed thread-safe). Use one client
```
Comment (DrPyser) [inline README.md]: probably not necessary to discuss this in length here. Quick description of the options and a quick mention of Keep-Alive/TCP connection reuse being enabled by default would be enought.

File: wazo_lib_rest_client/client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_lib_rest_client/client.py:115]: But at this point, would probably be preferable to keep the implementation simple and have clients be themselves thread-local?

[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_lib_rest_client/client.py:115]: https://github.com/psf/requests/issues/2766

File: wazo_lib_rest_client/client.py [hunk 6]
```diff
@@ -93,13 +94,33 @@ def _load_plugins(self) -> None:
...
+        requests across threads; use one client per thread in that case.
+        """
+        if self._session is None:
+            self._session = self._create_session()
+        return self._session
+
+    def _create_session(self) -> Session:
         session = Session()
-        session.headers = {'Connection': 'close'}
+        session.headers = {}
```
[the remark this answers is not in the document]
Comment (fblackburn1) [reply wazo_lib_rest_client/client.py]: There's some years since it has been fixed

File: wazo_lib_rest_client/client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fblackburn1) [reply wazo_lib_rest_client/client.py:115]: 780d610 adds a lock around session creation to fix the lazy-init race. The session stays shared across threads: urllib3's connection pool is thread-safe, and token/tenant/timeout are injected per request, so thread-local clients shouldn't be needed.

File: wazo_lib_rest_client/client.py (hunk 4 above)
Comment (fblackburn1) [reply wazo_lib_rest_client/client.py]: Trimmed in 780d610

**Comment by fblackburn1** [comment]
check experimental

**Comment by fblackburn1** [comment]
check experimental

**Comment by DrPyser** [comment]
won't do for now.
